flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java

```diff
@@ -140,6 +150,127 @@ public void testBlockingPartitionResetting() throws Exception {
         assertFalse(consumedPartitionGroup.areAllPartitionsFinished());
     }
 
+    @Test
+    public void testGetNumberOfSubpartitionsForNonDynamicAllToAllGraph() throws Exception {
+        testGetNumberOfSubpartitions(7, DistributionPattern.ALL_TO_ALL, false, Arrays.asList(7, 7));
+    }
+
+    @Test
+    public void testGetNumberOfSubpartitionsForNonDynamicPointWiseGraph() throws Exception {
+        testGetNumberOfSubpartitions(7, DistributionPattern.POINTWISE, false, Arrays.asList(4, 3));
+    }
+
+    @Test
+    public void testGetNumberOfSubpartitionsFromConsumerParallelismForDynamicAllToAllGraph()
+            throws Exception {
+        testGetNumberOfSubpartitions(7, DistributionPattern.ALL_TO_ALL, true, Arrays.asList(7, 7));
```

Comment:
Yes, it exists, users can set parallelism of a certain vertex. The adaptive batch scheduler only decide parallelism for vertex whose parallelim is not set by user(the vertex's parallelism in jobgraph is -1). For the job vertex whose parallelism is set by user or set in compile phase (the vertex's parallelism in jobgraph is > 0), adaptive batch scheduler will not change it's parallelism.

Reply:
This design is because the parallelism of some operators can only be a certain value. For example, `GlobalAgg`, it's parallelism can only be 1. In this case, the parallelism of `GlobalAgg` will be set to 1 in jobgraph during compilation, and the adaptive batch scheduler will not change it's parallelism.